Migrate cpumanager to contextual logging

ffromani opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Move to contextual logging,  which is an enabler to make the cpumanager logs easier to follow in the reconciliation and the cleanup flows.
This is part of an ongoing effort to make it easier to understand and troubleshoot the behavior of the resource managers.


#### Which issue(s) this PR fixes:
Fixes 123037 (partially, see issue)
Fixes 130069 (partially, see issue)

#### Special notes for your reviewer:
I'm contemplating alternative options about splitting this PR. I'm not sure targeting only parts of cpumanager (e.g. the flow mentioned in the description) is a better approach but I'm open to alternatives.

#### Does this PR introduce a user-facing change?
```release-note
Migrate the cpumanager to contextual logging
```

ffromani commented:
depends on #120569 

/wg structured-logging 
/area logging
/priority important-longterm
/kind cleanup
/cc @kubernetes/wg-structured-logging-reviews

(per https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#what-to-include-in-the-pull-request)

bart0sh commented on cmd/kubelet/app/server.go:
```diff
@@ -545,7 +545,7 @@ func getReservedCPUs(machineInfo *cadvisorapi.MachineInfo, cpus string) (cpuset.
 	}
 
-	topo, err := topology.Discover(machineInfo)
+	topo, err := topology.Discover(klog.Background(), machineInfo)
```
Would it make sense to pass context from the caller to this function instead of creating a new one?

pohly replied:
Yes. Typically, requiring a context or logger in some "leaf" function trickles all the way up to main (in a binary) or up to the `Test*` unit test function.

If you want to split up that work across different PRs, then use `klog.TODO()` as an indication that this isn't the final code.

ffromani replied:
it's simpler: I haven't noticed the parent context. Trivial overlook on my side. Will fix (and be more careful next time)

bart0sh commented on pkg/kubelet/cm/cpumanager/policy_none_test.go:
```diff
@@ -33,7 +40,10 @@ func TestNonePolicyName(t *testing.T) {
 
 func TestNonePolicyAllocate(t *testing.T) {
-	policy := &nonePolicy{}
+	logger, _ := ktesting.NewTestContext(t)
```
```suggestion
	logger, ctx := ktesting.NewTestContext(t)
```

ffromani replied:
thanks, fixed

bart0sh commented on pkg/kubelet/cm/cpumanager/policy_none_test.go:
```diff
@@ -43,14 +53,17 @@ func TestNonePolicyAllocate(t *testing.T) {
 
 	container := &testPod.Spec.Containers[0]
-	err := policy.Allocate(st, testPod, container)
+	err := policy.Allocate(context.Background(), st, testPod, container)
```
```suggestion
	err := policy.Allocate(ctx, st, testPod, container)
```

ffromani replied:
thnaks, fixed

mengjiao-liu commented:
/test pull-kubernetes-e2e-capz-windows-master

bart0sh commented on pkg/kubelet/cm/cpumanager/policy_static_test.go:
```diff
@@ -145,13 +149,14 @@ func TestStaticPolicyStart(t *testing.T) {
 				defaultCPUSet: testCase.stDefaultCPUSet,
 			}
-			err := policy.Start(st)
+			err := policy.Start(context.Background(), st)
```
Instead of creating new context with `context.Background()` you can use test context returned by the `ktesting.NewTestContext` above.

bart0sh replied:
oops, sorry. I've already suggested that before.

ffromani replied:
we can still improve it anyway, on it

bart0sh commented on cmd/kubelet/app/server.go:
```diff
@@ -749,7 +750,7 @@ func run(ctx context.Context, s *options.KubeletServer, kubeDeps *kubelet.Depend
 			return err
 		}
-		reservedSystemCPUs, err := getReservedCPUs(machineInfo, s.ReservedSystemCPUs)
+		reservedSystemCPUs, err := getReservedCPUs(klog.FromContext(ctx), machineInfo, s.ReservedSystemCPUs)
```
Although passing logger is acceptable, I'd suggest to consider passing a context. Context generally is more useful and can be used not just to get a logger.

bart0sh replied:
This is a generic comment about the PR, not just for this particular case.

ffromani replied:
makes sense, done

ffromani commented:
I\ll be resuming work on this once the preparation PR gets in (which should happen soon)

pohly commented:
@ffromani: still interested in this?

ffromani commented:
> @ffromani: still interested in this?

YES! I'm in touch with @swatisehgal to revamp this and also fixing memory manager

ffromani commented:
/reopen

ffromani committed:
node: cpumanager: migrate to contextual logging

migrate the cpumanager code to contextual logging

design musings:
- golang contexts "[...] carries deadlines, cancellation signals, and
  other request-scoped values across API boundaries and between
  processes." (see: https://pkg.go.dev/context#pkg-overview).
  Thus, it generally makes sense to add contexts to most of the APIs
  and functions  as the resource management is done in the admission flow, when
  ultimately we process a request and we have time bounds;
- however, receiving a context from the outside and depending on it
  kind also conflicts with the goal of having uniform logging and "
  - attach key/value pairs that get included in all log messages
  - add names that describe which component or operation triggered a log messages"
  (see:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging#summary)
  Some of this conflict is due to the transitional nature of the PRs
  which we do when updating the codelet codebase. We can envision a
  final state on which the top level caller adds identifying information
  like pod UID and pod namespace/name pair; but up until all the calling
  chain is updated and wired, the results are gonna be inconsistent.
- continuing from the point above, the most realistic transition plan
  for the kubelet is outlined in
  https://github.com/kubernetes/kubernetes/issues/130069 .
  Since we do want to move in parallel and migrate on a
  subsystem-by-subsystem basis, each subsystem will be in charge to add
  the data they want to their contexts. I think this is the only
  reasonnable way forward, and the cost will be overhead and redundancy.
  We should do sweeping changes to uniform the flows once we completed
  the full transition to ensure uniformity.
- there are some well known packages which want to operate by design
  only in in-memory data structure, which arguably can't block.
  Shoehorning a context feels not great. Instead, we intentionally pass
  only the logger, which still is the core idea of contextual logging
  (see:
https://github.com/kubernetes/kubernetes/pull/125912#discussion_r1935919144)
  examples are the topology subpackage and the cpu_assignment logic
  (takeByTopology and friends)

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
updated. I'm now thinking about how to make it easier to review (split in more commits).
Pending comments still TODO

ffromani commented:
/remove-lifecycle rotten

ffromani commented:
/retest

bart0sh commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -320,6 +320,7 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
 
 	// Initialize CPU manager
 	cm.cpuManager, err = cpumanager.NewManager(
+		klog.Background(),
```
`NewContainerManager` is called from the code that has a context. Would it make sense to pass it to the `NewContainerManager ` and use here instead of klog.Background() ?

ffromani replied:
Yes, this is top level wiring we can do, but I will do in a different commit in this PR

bart0sh commented on pkg/kubelet/cm/container_manager_stub.go:
```diff
@@ -125,7 +125,7 @@ func (cm *containerManagerStub) UpdatePluginResources(*schedulerframework.NodeIn
 
 func (cm *containerManagerStub) InternalContainerLifecycle() InternalContainerLifecycle {
-	return &internalContainerLifecycleImpl{cpumanager.NewFakeManager(), memorymanager.NewFakeManager(), topologymanager.NewFakeManager()}
+	return &internalContainerLifecycleImpl{cpumanager.NewFakeManager(klog.Background()), memorymanager.NewFakeManager(), topologymanager.NewFakeManager()}
```
The same here.

ffromani replied:
Here I think we can just save the context being passed and reuse here. It's functionally equivalent.

ffromani commented:
PS: @bart0sh I still have pending comments to address, I'll do and then do another upload

bart0sh commented on pkg/kubelet/cm/container_manager_stub.go:
```diff
@@ -41,14 +41,16 @@ import (
-func (cm *containerManagerStub) Start(_ context.Context, _ *v1.Node, _ ActivePodsFunc, _ GetNodeFunc, _ config.SourcesReady, _ status.PodStatusProvider, _ internalapi.RuntimeService, _ bool) error {
+func (cm *containerManagerStub) Start(ctx context.Context, _ *v1.Node, _ ActivePodsFunc, _ GetNodeFunc, _ config.SourcesReady, _ status.PodStatusProvider, _ internalapi.RuntimeService, _ bool) error {
 	klog.V(2).InfoS("Starting stub container manager")
+	cm.ctx = ctx
```
[Context documentation](https://pkg.go.dev/context) says that storing context should be avoided:
```
Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx.
```

[The Go blog](https://go.dev/blog/context-and-structs) elaborates on this quite nicely, PTAL.

ffromani replied:
Nice read, thanks. There's an argument to be made ineed about the tradeoff for the stub, preserving the backward compatibility vs making the currently trivial (and beneficially trivial) stub code simple vs adding more complexity. So I pivoted a bit more in the latest upload.

ffromani replied:
Basically, the friction point is that adding the context to the `InternalContainerLifecycle` call is not optimal: this will be needed just and only in the `stub code`, the real impls don't need it.
So it boils down to what to store in the stub object and what not.

bart0sh commented on pkg/kubelet/cm/cpumanager/cpu_assignment_test.go:
```diff
@@ -214,7 +218,8 @@ func TestCPUAccumulatorFreeNUMANodes(t *testing.T) {
 	for _, tc := range testCases {
 		t.Run(tc.description, func(t *testing.T) {
-			acc := newCPUAccumulator(tc.topo, tc.availableCPUs, 0, CPUSortingStrategyPacked)
+			logger, _ := ktesting.NewTestContext(t)
```
I think `t` is the same here as at the start of `TestCPUAccumulatorFreeNUMANodes`, so it should be enough to call `ktesting.NewTestContext(t)` there once.

ffromani replied:
good point, fixed and fixed all the similar instances

bart0sh commented:
@ffromani Can you add `pkg/kubelet/cm/cpumanager` to `https://github.com/kubernetes/kubernetes/blob/master/hack/logcheck.conf` and generate `golangci` yamls with `hack/update-golangci-lint-config.sh`? This way cpumanager code will be checked automatically by some ci job.

ffromani commented:
> @ffromani Can you add `pkg/kubelet/cm/cpumanager` to `https://github.com/kubernetes/kubernetes/blob/master/hack/logcheck.conf` and generate `golangci` yamls with `hack/update-golangci-lint-config.sh`? This way cpumanager code will be checked automatically by some ci job.

absolutely, good idea. Adding for the next upload (this PR is already on hold for reasons explained above)

ffromani commented:
/test pull-kubernetes-e2e-gce

ffromani commented:
/hold cancel

ffromani committed:
cpumanager: hack: add golang ci config for contextual logging

add cpumanager to contextual logging linter
regenerate with
```hack/update-golangci-lint-config.sh```

Signed-off-by: Francesco Romani <fromani@redhat.com>

swatisehgal commented:
> @ffromani Can you add `pkg/kubelet/cm/cpumanager` to `https://github.com/kubernetes/kubernetes/blob/master/hack/logcheck.conf` and generate `golangci` yamls with `hack/update-golangci-lint-config.sh`? This way cpumanager code will be checked automatically by some ci job.

Considering we want https://github.com/kubernetes/kubernetes/pull/129685 to get merged first. Does it make sense to add `cpumanager` to logcheck file in that PR?

swatisehgal commented:
> @swatisehgal
> 
> > Considering we want #129685 to get merged first. Does it make sense to add cpumanager to logcheck file in that PR?
> 
> I don't think it will work as #129685 doesn't change logging to contextual. It's still structured logging and all kubelet codebase has been already moved to structured logging. That's why there is a line `structured k8s.io/kubernetes/pkg/kubelet/.*` in logcheck.conf.

I see, Thanks for clarifying.

swatisehgal commented on pkg/kubelet/cm/cpumanager/cpu_assignment_test.go:
```diff
@@ -21,11 +21,15 @@ import (
 )
 
 func TestCPUAccumulatorFreeSockets(t *testing.T) {
+	logger, _ := ktesting.NewTestContext(t)
```
I have seen the use of `ktesting.Init(t)` for creation of logger for testing. Do we know why we use `NewTestContext(t)` vs `Init(t) `?

bart0sh replied:
`ktesting.Init(t)` returns only a context and `ktesting.NewTestContext(t)` [calls `ktesting.Init(t)`](https://github.com/kubernetes/kubernetes/blob/master/test/utils/ktesting/ktesting.go#L45-L48) **and** returns 
 a context and a logger. As we need a logger here, it's shorter to call `ktesting.Init(t)`

ffromani commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -369,18 +378,18 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 
 	// Allocate CPUs according to the NUMA affinity contained in the hint.
-	cpuset, err := p.allocateCPUs(s, numCPUs, hint.NUMANodeAffinity, p.cpusToReuse[string(pod.UID)])
+	cpuset, err := p.allocateCPUs(lh, s, numCPUs, hint.NUMANodeAffinity, p.cpusToReuse[string(pod.UID)])
```
I would like to reuse the `loggerWithValues` I created on line 328 above to carry the context. I can't recreate (which would be a little wasteful, but OK) without passing parameters, and these would be used only to build a logger, which seems not great to me.

Another option is to create another context on the fly to wrap just the logger, but `allocateCPUs` and all its callees only do memory operations, so trivial to interrupt and (hopefully? :) ) always very fast, so there's little justification for passing down a context IMO, especially considering it would only wrap a logger instance.

So for these reasons I'm only passing a logger, which is exactly what we need, and not a full context.
However if the guidelines and/or the reviewers prefer otherwise I can change to add the context, no probs.

pohly replied:
Passing a logger in situations like this is perfectly fine.

"Contextual logging" gets its name from "the context of a function call determines the logger", not from "always get a logger from a `context.Context`".

ffromani commented on pkg/kubelet/cm/container_manager_windows.go:
```diff
@@ -134,8 +134,10 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
+
 	cm.topologyManager = topologymanager.NewFakeManager()
-	cm.cpuManager = cpumanager.NewFakeManager()
+	cm.cpuManager = cpumanager.NewFakeManager(logger)
```
ok, took a bit of extra changes, but done

ffromani commented:
yuck, still known issue. My fault. Running some more checks locally before to upload again (this time for good I hope!)

swatisehgal commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -203,7 +203,7 @@ func validateSystemRequirements(mountUtil mount.Interface) (features, error) {
 // Takes the absolute name of the specified containers.
 // Empty container name disables use of the specified container.
-func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.Interface, nodeConfig NodeConfig, failSwapOn bool, recorder record.EventRecorder, kubeClient clientset.Interface) (ContainerManager, error) {
+func NewContainerManager(ctx context.Context, mountUtil mount.Interface, cadvisorInterface cadvisor.Interface, nodeConfig NodeConfig, failSwapOn bool, recorder record.EventRecorder, kubeClient clientset.Interface) (ContainerManager, error) {
```
Still see references to `klog.InfoS` in this function, think we should transition to  `logger.Info` as per https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#contextual-logging-in-kubernetes.

bart0sh replied:
I guess this is because this code doesn't belong to cpumanager. It should be changed when migrating container manager.

ffromani replied:
@bart0sh exactly. I'm changing this file anyway because I need to pass down the context. The minimal change would be to still change this function and pass down a `context.TODO()`. I'll review and update the PR ASAP.

I can help migrating `k/k/pkg/kubelet/cm` later on if the effort isn't started already

bart0sh replied:
Yep, this is exactly the approach I used in https://github.com/kubernetes/kubernetes/pull/129942 and outlined in https://github.com/kubernetes/kubernetes/issues/130069. I think it makes sense to continue with it.

swatisehgal replied:
Thanks for clarifying, makes sense!

swatisehgal commented on pkg/kubelet/cm/container_manager_stub.go:
```diff
@@ -43,12 +43,14 @@ import (
-func (cm *containerManagerStub) Start(_ context.Context, _ *v1.Node, _ ActivePodsFunc, _ GetNodeFunc, _ config.SourcesReady, _ status.PodStatusProvider, _ internalapi.RuntimeService, _ bool) error {
+func (cm *containerManagerStub) Start(ctx context.Context, _ *v1.Node, _ ActivePodsFunc, _ GetNodeFunc, _ config.SourcesReady, _ status.PodStatusProvider, _ internalapi.RuntimeService, _ bool) error {
 	klog.V(2).InfoS("Starting stub container manager")
+	cm.cpuManager = cpumanager.NewFakeManager(ctx)
```
Are we intentionally leaving  `klog.InfoS` and not updating to the logger equivalent in the line above? If so, why?

ffromani replied:
Yes, same reasoning per https://github.com/kubernetes/kubernetes/pull/125912#discussion_r1950480214 . I'm trying to do the minimal changes and move gradually once the deps are settled. This will cause some churn, but it's the only way I can think of to keep PRs size manageable.

swatisehgal commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -156,26 +158,29 @@ func (s *sourcesReadyStub) AddSource(source string) {}
 	var err error
 
-	topo, err = topology.Discover(machineInfo)
+	logger := klog.LoggerWithName(klog.FromContext(ctx), "cpu")
```
Just curious: Is it okay to not have key value pair in `LoggerWithName`?

ffromani replied:
Yes, for `WithName` yes:
```
// LoggerWithName returns logger.WithName(name) when contextual logging is
// enabled, otherwise the logger.
func LoggerWithName(logger Logger, name string) Logger {
...
```

swatisehgal commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -378,65 +394,74 @@ func (m *manager) removeStaleState() {
-			klog.V(5).InfoS("RemoveStaleState: containerMap: container still active", "podUID", podUID, "containerName", containerName)
+			lh.V(5).Info("containerMap: container still active")
 			return
 		}
-		klog.V(2).InfoS("RemoveStaleState: containerMap: removing container", "podUID", podUID, "containerName", containerName)
-		err := m.policyRemoveContainerByRef(podUID, containerName)
+		lh.V(2).Info("containerMap: removing container")
+		err := m.policyRemoveContainerByRef(ctx, podUID, containerName)
```
Perhaps this could be a nice place to toy around with LoggerV and have log entries based on log verbosity (https://kubernetes.slack.com/archives/C20HH14P7/p1737563609490089)?

swatisehgal commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -346,14 +360,16 @@ type reconciledContainer struct {
 		return
 	}
 
+	lh_ := klog.FromContext(ctx)
```
Why the underscore in `lh_`? Is this a convention? Where is this documented?

ffromani replied:
In this function I'm repeatedly replacing the logger instance using shadowing (lacking any better alternative) in the loop body. I'm keeping a reference of the original passed logger to correctly recreate at each loop iteration, otherwise we unnecessarily grow the context each time. This way each iteration creates a new context with a fixed size payload.

The trailing underscore is the cheapest way to track the original variable: IMO `lhOrig` or `origLogHandle` are unnecessarily long and not really easier to understand.

I'm using the underscore name for the instance which is used less often: we use the function parameter only here.

swatisehgal commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -450,7 +475,7 @@ func (m *manager) reconcileState() (success []reconciledContainer, failure []rec
 				_, _, err := m.containerMap.GetContainerRef(containerID)
 				if err == nil {
-					klog.V(4).InfoS("ReconcileState: ignoring terminated container", "pod", klog.KObj(pod), "containerID", containerID)
+					lh.V(4).Info("ignoring terminated container", "containerID", containerID)
```
Why not update the `logger` with `containerID` and use that for logging below?

ffromani replied:
yep, you have it in the next comment

swatisehgal commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -465,17 +490,17 @@ func (m *manager) reconcileState() (success []reconciledContainer, failure []rec
 			if cset.IsEmpty() {
 				// NOTE: This should not happen outside of tests.
-				klog.V(2).InfoS("ReconcileState: skipping container; assigned cpuset is empty", "pod", klog.KObj(pod), "containerName", container.Name)
+				lh.V(2).Info("skipping container; assigned cpuset is empty")
```
Ah, I see because we don't have `containerID` here?

With a lower verbosity makes sense to skip this entry.

Another candidate for `loggerV`?

ffromani replied:
yes, I'll need to play a bit with `loggerV` to see how we can capture `containerID`. Will do for the next upload

swatisehgal commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -129,7 +133,9 @@ var _ Policy = &staticPolicy{}
 // container process starts.
-func NewStaticPolicy(topology *topology.CPUTopology, numReservedCPUs int, reservedCPUs cpuset.CPUSet, affinity topologymanager.Store, cpuPolicyOptions map[string]string) (Policy, error) {
+func NewStaticPolicy(ctx context.Context, topology *topology.CPUTopology, numReservedCPUs int, reservedCPUs cpuset.CPUSet, affinity topologymanager.Store, cpuPolicyOptions map[string]string) (Policy, error) {
+	logger := klog.LoggerWithName(klog.FromContext(ctx), "static")
```
Should we add the "policy" key here?
```suggestion
logger := klog.LoggerWithName(klog.FromContext(ctx), "policy", "static")
```

swatisehgal reviewed (commented):
Leaving some comments, will do another pass of this later.

ffromani commented:
/hold

1. still playing with `loggerV`
2. want to collect some logs to demonstrate how it looks

other than that I think I addressed all pending comments.

ffromani commented:
/retest-required

ffromani commented:
/retest

ffromani commented:
need to rebase, still unhappy with some internal details about how managing the logger, still playing wiht `loggerV`. Updating ASAP nevertheless

ffromani commented:
> need to rebase, still unhappy with some internal details about how managing the logger, still playing wiht `loggerV`. Updating ASAP nevertheless

pending items

- should `cpumanager.Manager` interface accepts contexts? Didn't make my mind yes but leaning towards yes
- still pending experiment with `loggerV`, next on my TODO list

since last upload:

- intentionally avoid to store `logr.Logger` instance as default. In some key and documented cases we better do, in most cases we should not and just always fetch `FromContext`
- cleanup in test cases

swatisehgal reviewed (commented):
I've started exploring the migration of the Topology Manager to contextual logging and noticed significant overlap in the higher-level plumbing of context and changes needed at higher level component e.g. container manager.

We can obviously identify the order of merging (which will typically be in the order the PRs were pushed) and then build on top of ongoing PRs. IOW topology manager contextual logging work would build on top of this work and base changes on existing PR. This approach is totally doable but could become messy when the work is going on in parallel. 

Alternatively, would it make sense to introduce separate commit(s) for the Container Manager (or other higher-level components) that apply to all resource managers? This would reduce duplication, improve code reuse across resource managers, simplify rebases when contributing, and allow us to work on all resource manager migrations in parallel rather than sequentially. WDYT?

bart0sh commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -257,6 +257,13 @@ const (
+	// `takeByTopology` call. Hence there's no functional difference between storing it or
+	// getting repeatedly from the calls. Furthermore, this way we can keep the API surface smaller
+	// (less parameters).
+	logger logr.Logger
```
I'd suggest to use klog.Logger instead. logr is an indirect dependency and theoretically can be changed.

ffromani replied:
sure, will change in the next upload

bart0sh commented:
/reopen

ffromani commented:
/remove-lifecycle rotten

ffromani commented:
/hold cancel

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

ffromani commented:
not sure why the lane failed only now, but the issue is real, I missed to update a couple places. Fixed now.

swatisehgal reviewed (commented):
/lgtm
